### PR TITLE
Add some test flags for e2e test

### PR DESCRIPTION
**Pull request by ypu**
In this pullreq, mainly doing one thing by add a command line flags.
1. Run e2e test with an existing or old version of etcd and etcdctl.
The flag name is 'bin-dir' and it is default value is "../bin" as it is in the original code.

Example command line(run under e2e directory):
 go test -timeout 10m -v -bin-dir /usr/bin

### CODE REVIEW

**Comment by heyitsanthony** [comment]
> 1. Run e2e test with an existing or old version of etcd and etcdctl.
Why does this need special support? Just put/symlink the old version to bin/ and run the e2e tests as usual.

> 2. Run e2e tests based on its category.
This should probably be done by changing the names of the tests so they have prefixes like `TestV2Whatever`, `TestV3Whatever` so that `go test -run=TestV2` will run all the V2 tests.

**Comment by ypu** [comment]
@heyitsanthony Thanks a lot for your comments. Forget that we can set a pattern when running the test. So drop the commit for it.

But I think it is worth to add the flag to make it easier to run with a custom binary. For example if I have several hosts with etcd package installed. And I want to run the e2e test in them but I don't want to deploy the whole go src and etcd src env on them. I just need compile the e2e package with go test -c ./e2e and then copy the e2e.test to the hosts and run it with -bin-dir=/usr/bin. With this flag's help we don't need to make a separated dir with request structure, or put it under /usr/bin/$dir_for_etcd_e2e/ in the system. Also when we have more than one binary in the host, this also will help to test them one by one quickly without any setup changes..

**Comment by heyitsanthony** [comment]
@ypu OK, but what's the point of running only the e2e tests instead of all the tests? It's not the complete test suite and ignores a lot of the system functionality.

**Comment by ypu** [comment]
@heyitsanthony The other tests provided such like fmt check, unit test, build tests are all related with the source code. But end to end test is different. It tests the binary behaviour but not the source code.  So maybe it is meaningful to make it can run with any binary we want to test.

It is also very helpful to test etcd and etcdctl in an product env, such as systems have kubernetes and so on. In such env people may not have the source code for their binaries, but they can use e2e test for the functional testing.

**Comment by heyitsanthony** [comment]
@ypu No. The unit and integration tests execute code paths never covered by the e2e tests. They're not static checks like the fmt or build passes; the code is compiled and runs. e2e tests run only a very limited amount of the code in the full binary; they are not sufficient for testing the entire system and were, for the most part, only intended to ensure that basic command line functionality works.

File: e2e/etcd_test.go [hunk 1]
```diff
@@ -433,9 +433,9 @@ func (ep *etcdProcess) Stop() error {
 }
 
 func (ep *etcdProcess) waitReady() error {
-	readyStrs := []string{"enabled capabilities for version", "published"}
+	readyStrs := []string{"starting server...", "published"}
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: this line is important; otherwise the processes aren't guaranteed to negotiate up to the 3.0 api by the time the test runs

**Comment by heyitsanthony** [comment]
I guess this change is minimal enough to be OK once CI passes but I think it needs #5088 to make much sense in terms of testing. @xiang90 any thoughts on supporting this use case?

File: e2e/etcd_test.go (hunk 1 above)
Comment (ypu) [inline e2e/etcd_test.go]: Yes, noticed that from the past commits that added this check. But  maybe it is not suitable to check this in waitReady function. This function should only check the server is fully up. But this check will fail the tests in old version of etcd such as 2.3.7. This should not be a must check point in v2 related tests. Maybe it is better to add this check only in v3 related tests?

Comment (heyitsanthony) [inline e2e/etcd_test.go]: No, that's not going to happen. The tests are meant for the current version of etcd; tests are only backported when it's convenient. It's too much trouble to worry about whether changes to the current tests break on older versions of etcd and there's not enough CI resources to support it. Are there any go projects with test suites which are backwards compatible like that?

Comment (ypu) [inline e2e/etcd_test.go]: As I know that the kubernetes e2e can run with a running cluster. The cluster may have a lower version than the code. Not sure they plan to do this or not. But seems it works.
In this pullreq I am trying to do the same thing for etcd e2e test. And for a basic function like waitReady maybe it is better to use some more common string for checking. And the request of checking the process is guaranteed to negotiate up to the 3.0 api or not should be in the tests or other functions.

But it is OK for me to keep it the original way as this is not the main purpose of the pullreq.

Comment (heyitsanthony) [inline e2e/etcd_test.go]: OK, I see kubernetes has full-fledged versioned e2e conformance tests; that versioning functionality is definitely intentional. Forward rolling testing (e.g., use 3.0.x tests against 3.1) seems reasonable for e2e, integration, and clientv3/integration tests using etcd binaries independent from test fixtures, but it'll still need some thought to get it right.

**Comment by xiang90** [comment]
@heyitsanthony I am not sure how this can benefit Kubernetes. For Kuberenetes, we probably want to rely on Kubernetes testing to ensure the etcd binary is compatible... I do not quite understand how etcd e2e test can help with that. But I am fine with simply making the test infra more flexible with adding little code.

File: e2e/etcd_test.go (hunk 1 above)
Comment (ypu) [inline e2e/etcd_test.go]: OK. Will drop this commit in this pullreq and submit a separated one for this issue. Thanks a lot for your comments.

**Comment by ypu** [comment]
@heyitsanthony and @xiang90 drop the commit for update the string in waitReady, will use another pullreq for it. And currently this pullreq only about make e2e specific the binary and cert files directories  from command line. Can you help to review it again?

**Comment by heyitsanthony** [comment]
lgtm

**Pull request merged**